docs/dense_set.md

```diff
@@ -0,0 +1,32 @@
+# DenseSet in Dragonfly
+
+`DenseSet` uses [classic hashtable with separate chaining](https://en.wikipedia.org/wiki/Hash_table#Separate_chaining) similar to the Redis dictionary for lookup of items within the set.
+
+The main optimization present in `DenseSet` is the ability for a pointer to **point to either an object or a link key**, removing the need to allocate a set entry for every node. This is accomplished by using [pointer tagging](https://en.wikipedia.org/wiki/Tagged_pointer) exploiting the fact that the top 12 bits of any userspace address are not used and can be set to indicate if the current pointer points to nothing, a link key, or an object.
+
+Further, to reduce collisions items may be inserted into neighbouring cells that are empty to reduce the number of unused spaces. These nodes are then marked as displaced using pointer tagging.
+
+An example of possible bucket configurations can be seen below. Note the bit tagging shown in each pointer, indicating the type of object the pointer references and if it is displaced.
+
+In order from top to bottom the following types of chains are shown:
```

Comment:
hah, I missed this explanation when looking at the diagram. Maybe it's better to add text boxes under each example to make it self contained?